### DISCUSSION TITLE
Nested ViewModels and Models?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16280] [Q&A]
Say I have some sort of class that represents a game save or a level format..
How would I setup it so nested data can be accessed in the viewmodel from the view?

For example a level might have several rooms with several layers each containing different types of objects that hold onto sprite data.

Or how about one viewmodel with two views? maybe one for viewing and one for editing (maybe a toolbar that selects sections).

**Comment by thevortexcloud** [comment]
If you are dealing with unknown levels of nesting with unknown types, you can use a tree view (and also possibly a view locator) as they support recursive hierarchy data. Avalonia will do a lot of the hard work for you in that case. You just have to load it into the view somehow. EG put everything into an observable collection and a tree view. 

https://docs.avaloniaui.net/docs/reference/controls/treeview-1

The tree datagrid can also be used and supports some different scenarios.

https://docs.avaloniaui.net/docs/reference/controls/treedatagrid/
https://docs.avaloniaui.net/docs/reference/controls/treedatagrid/creating-a-hierarchical-treedatagrid

If you are just dealing with a flat list, a simple observable collection and items control will work fine.

> Or how about one viewmodel with two views? 

Technically the view model would not contain the view if you are properly following MVVM. But assuming you mean your view model contains a fixed number of child view models, this is fairly easy. Just do something like

```xaml
<ContentControl Content="{CompiledBinding ChildViewModel}">
        <ContentControl.DataTemplates>
        <!--Define template for ChildViewModel here!-->
        </ContentControl.DataTemplates>
</ContentControl>
```

> maybe one for viewing and one for editing (maybe a toolbar that selects sections).

There are several ways this could be implemented. I typically just do it as a single view to reduce duplication, and just have a mode property on the view model. You can then just create a converter that will enable/disable things on the view based on the current mode state. But if you want to have different views you can either:

1. Manually create the views in the code behind (EG when a button is pressed) and reuse the existing view model
2. Create a dedicated edit/view version of the view model and define a template for each

Depending on your actual goals though will probably influence how you deal with this.
